### PR TITLE
Adding payload and decoded tabs per transaction.

### PR DESCRIPTION
**Pull request by hbulgarini**
This PR adds the payload and decoded tabs per transaction detailed in #200.

Also i have included the json display proposal discussed on https://github.com/paritytech/parity-bridges-ui/issues/200#issuecomment-869690103 . In case @goldsteinsveta  or @tomusdrw do not agree with the proposal this PR can be cancelled and rebuilt accordingly. IMHO , the proposal adds value to the payload inspection and in case the raw payload is needed, the copy feature can be used.

EDIT: `semantic-ui-react` package was finally removed from the project. Related #67 .

### CODE REVIEW

**Review by wirednkod — COMMENTED** [review]
Overall looks good - but the result is not as per #200 screenshots. There are details missing (suvh as icons and/or titles per tab)

File: src/components/TransactionSwitchTab.tsx [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+const TransactionSwitchTab = ({ transactionDisplayProps, children, payloadHex, TransactionDisplayPayload }: Props) => {
+  const [tab, setTab] = useState(SwitchTabType.RECEIPT);
+
+  const getColorByState = (value: string) => (value === tab ? 'primary' : 'secondary');
+
+  return (
+    <>
+      <Box mt={2}>
+        <ButtonSwitchMode color={getColorByState(SwitchTabType.PAYLOAD)} onClick={() => setTab(SwitchTabType.PAYLOAD)}>
```
Comment (tomusdrw) [inline src/components/TransactionSwitchTab.tsx:42]: Every `onClick` should be `useCallback` really, but I can turn my blind eye on this.

File: src/hooks/transactions/useTransactionPayloadDisplay.ts [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    if (!receiverAddress) {
+      setPayloadHex(null);
+      setTransactionDisplayPayload(null);
+    }
+    if (payload) {
+      //@ts-ignore
+      const payloadType = createType(sourceChain, 'OutboundPayload', payload);
+      setPayloadHex(payloadType.toHex());
+    }
+    //@ts-ignorex
```
Comment (tomusdrw) [inline src/hooks/transactions/useTransactionPayloadDisplay.ts:45]:
```suggestion
    //@ts-ignore
```

File: src/util/transactionUtils.ts [hunk 3]
```diff
@@ -14,8 +14,39 @@
...
+    TransactionDisplayPayload.origin = {
+      SourceAccount: toObject(origin.SourceAccount)
+    };
+
+    TransactionDisplayPayload.weight = weight;
+    TransactionDisplayPayload.spec_version = spec_version;
+    return TransactionDisplayPayload;
+  }
+  return null;
+}
```
Comment (tomusdrw) [inline src/util/transactionUtils.ts:52]:
```suggestion
export function getTransactionDisplayPayload(payload: Payload): TransactionDisplayPayload {
  const transactionDisplayPayload = {} as TransactionDisplayPayload;
  const { call, origin, spec_version, weight } = payload;
  transactionDisplayPayload.call = toObject(call);
  transactionDisplayPayload.origin = {
    SourceAccount: toObject(origin.SourceAccount)
  };

  transactionDisplayPayload.weight = weight;
  transactionDisplayPayload.spec_version = spec_version;
  return transactionDisplayPayload;
}
```

1. Keep the `null` check outside of the function (you're doing it anyway on the call site).
2. Use lower-case names for variables 😱

File: src/hooks/transactions/useTransactionPayloadDisplay.ts [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+  useEffect(() => {
+    if (!receiverAddress) {
+      setPayloadHex(null);
+      setTransactionDisplayPayload(null);
+    }
+    if (payload) {
+      //@ts-ignore
+      const payloadType = createType(sourceChain, 'OutboundPayload', payload);
+      setPayloadHex(payloadType.toHex());
```
Comment (tomusdrw) [inline src/hooks/transactions/useTransactionPayloadDisplay.ts:43]: Incorrect. If `payload==null` the old hex will be left in state.

File: src/types/transactionTypes.ts [hunk 5]
```diff
@@ -65,6 +78,8 @@ export interface TransactionState {
   payload: TransactionPayload | null;
   payloadError: string | null;
   estimatedFeeLoading: boolean;
+  payloadHex: string | null;
```
Comment (tomusdrw) [inline src/types/transactionTypes.ts:81]: group with other `payload` types.

File: src/types/transactionTypes.ts [hunk 6]
```diff
@@ -65,6 +78,8 @@ export interface TransactionState {
...
+  TransactionDisplayPayload: TransactionDisplayPayload;
```
Comment (tomusdrw) [inline src/types/transactionTypes.ts:82]: lowercase
```suggestion
  transactionDisplayPayload: TransactionDisplayPayload;
```

File: src/util/transactionUtils.ts [hunk 7]
```diff
@@ -14,8 +14,39 @@
...
+  TransactionDisplayPayload
+} from '../types/transactionTypes';
 
 export function isTransactionCompleted(transaction: TransactionStatusType): boolean {
   return transaction.status === TransactionStatusEnum.COMPLETED;
 }
+
+interface ToObject {
+  [propName: number]: number;
+}
```
Comment (tomusdrw) [inline src/util/transactionUtils.ts:30]: Why is this needed at all?

File: src/util/transactionUtils.ts [hunk 8]
```diff
@@ -14,8 +14,39 @@
...
+
+function toObject(arr: Array<any>): Object {
+  const rv: ToObject = {};
+  for (let i = 0; i < arr.length; ++i) rv[i] = arr[i];
+  return rv;
+}
```
Comment (tomusdrw) [inline src/util/transactionUtils.ts:36]:
```suggestion
function toObject(arr: Array<any>): Object {
 return Object.fromEntries(Object.entries(arr));
}
```

I'd suggest to make it internal to `getTransactionDisplayPayload` - with shortr implementation we can even inline this.

**Review by tomusdrw — COMMENTED** [review]
Can you please add a screenshot of what it looks like?

**Comment by hbulgarini** [comment]
> #200

Yes! i need to add the header thanks!

**Comment by hbulgarini** [comment]
> Can you please add a screenshot of what it looks like?

Screenshots are attached [here](https://github.com/paritytech/parity-bridges-ui/issues/200#issuecomment-869690103), do you need additional screenshots?

File: src/util/transactionUtils.ts (hunk 7 above)
Comment (hbulgarini) [reply src/util/transactionUtils.ts:30]: it is not needed anymore after your proposal.

File: src/util/transactionUtils.ts (hunk 3 above)
Comment (hbulgarini) [reply src/util/transactionUtils.ts:52]: 1. OK
2. Daaammmnnn 😱 , that was a mistake that happened after a replace all and i didn't realize

File: src/hooks/transactions/useTransactionPayloadDisplay.ts (hunk 2 above)
Comment (hbulgarini) [reply src/hooks/transactions/useTransactionPayloadDisplay.ts:45]: i don't know why this is even here.

File: src/hooks/transactions/useTransactionPayloadDisplay.ts (hunk 4 above)
Comment (hbulgarini) [reply src/hooks/transactions/useTransactionPayloadDisplay.ts:43]: correct.

File: src/components/TransactionSwitchTab.tsx (hunk 1 above)
Comment (hbulgarini) [reply src/components/TransactionSwitchTab.tsx:42]: no. it should be as you mention.

**Comment by tomusdrw** [comment]
> Screenshots are attached here, do you need additional screenshots?

Yes. I'd like to see how it's integrated with the current UI, not only the payload part.

File: src/components/Transactions.tsx [hunk 9]
```diff
@@ -34,6 +35,7 @@ const Transactions = ({ type, ...transactionDisplayProps }: Props) => {
   const { transactions } = useTransactionContext();
   const { dispatchTransaction } = useUpdateTransactionContext();
   const { dispatchMessage } = useUpdateMessageContext();
+  useResetTransactionState(type);
```
Comment (tomusdrw) [inline src/components/Transactions.tsx:38]: Can you elaborate what this is about? What is `type` and why if we set it back to a previous value it should reset the transaction state?

File: src/contexts/TransactionContext.tsx [hunk 10]
```diff
@@ -57,7 +57,9 @@ export function TransactionContextProvider(props: TransactionContextProviderProp
     payloadError: null,
     showBalance: false,
     formatFound: null,
-    estimatedFeeLoading: false
+    estimatedFeeLoading: false,
+    payloadHex: null,
```
Comment (tomusdrw) [inline src/contexts/TransactionContext.tsx:61]: Sort all props please or at least move related stuff next to each other :)

File: src/hooks/transactions/useTransactionPayloadDisplay.ts [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+
+import { TransactionDisplayPayload } from '../../types/transactionTypes';
+import { getTransactionDisplayPayload } from '../../util/transactionUtils';
+import useApiCalls from '../api/useApiCalls';
+import { useAccountContext } from '../../contexts/AccountContextProvider';
+
+const useTransactionPayloadDisplay = () => {
+  const { payload } = useTransactionContext();
+  const [payloadHex, setPayloadHex] = useState<string | null>(null);
+  const [transactionDisplayPayload, setTransactionDisplayPayload] = useState<TransactionDisplayPayload | null>(null);
```
Comment (tomusdrw) [inline src/hooks/transactions/useTransactionPayloadDisplay.ts:29]: I feel these two states could be turned into one. We change them in unison anyway.

File: src/util/transactionUtils.ts [hunk 12]
```diff
@@ -14,8 +14,46 @@
...
+  payload,
+  ss58Format,
+  account,
+  createType,
+  targetChain
+}: PayloadInput): TransactionDisplayPayload {
+  //@ts-ignore
+  const callType = createType(targetChain, 'BridgedOpaqueCall', payload.call);
+  //@ts-ignore
+  const call = createType(targetChain, 'Call', callType.toHex());
```
Comment (tomusdrw) [inline src/util/transactionUtils.ts:47]:
```suggestion
  const call = createType(targetChain, 'Call', payload.call);
```
I think this should work just fine, no?

**Review by tomusdrw — COMMENTED** [review]
Awesome stuff! Thanks for the screenshots.

I think ideally if we could convert `callIndex` to an actual name of the pallet and dispatchable (`balances::transfer`), but let's make it a stretch goal in a separate PR :)

I also feel there is a bit too much happening in the JSON-display component. Is it styleabe? I'd honestly hide a bunch of things, like:
- number of items
- types
- "toggle triangles" (dunno the professional name :)) could be made lighter
- the vertical guides could be removed

CC @goldsteinsveta wdyt?

**Comment by hbulgarini** [comment]
@goldsteinsveta  i had to do a couple of UX change in order to be able to show a the payload of a second transaction that was not confirmed yet:  The`TransactionStatusMocked`component ([code](https://github.com/paritytech/parity-bridges-ui/blob/hb-add-payload-decoded-tab/src/components/Transactions.tsx#L42))  used to get unmounted after the first transaction so as a result the payload currently being built could not be checked until the second transaction was confirmed. I thought that the easiest solution until we fully implement #108.

Also once the transaction starts executing the receiver form gets completed reseted, which i think for now is the best.
Again in #108 the idea would be to lock the form .

File: src/util/transactionUtils.ts (hunk 12 above)
Comment (hbulgarini) [reply src/util/transactionUtils.ts:47]: i thought the same, but the api complained when it tried to built the type in that way.

File: src/components/Transactions.tsx (hunk 9 above)
Comment (hbulgarini) [reply src/components/Transactions.tsx:38]: Of course ... basically when i switch actions tabs ... let's say from "Transfer" to "CustomCall" the payload and estimatedFee needs to be reseted otherwise the tabs would show the values calculated in the previous tab.

Anyway it is not working 100% as expected since for this specific scenario i have an issue:

1) Fill the form for a Transaction tab.
2) Switch action to Custom Call

Estimated Fee remains the same as the one calculated from the previous tab. This is only happen for this scenario, the other switch do not face this issue. As i considered that this issue was out of the scope of this PR i have logged the following issues in order to track this:

https://github.com/paritytech/parity-bridges-ui/issues/195
https://github.com/paritytech/parity-bridges-ui/issues/209
https://github.com/paritytech/parity-bridges-ui/issues/207

File: src/contexts/TransactionContext.tsx (hunk 10 above)
Comment (hbulgarini) [reply src/contexts/TransactionContext.tsx:61]: sure ... i will also do it on the `transactionTypes`

File: src/util/transactionUtils.ts (hunk 12 above)
Comment (tomusdrw) [reply src/util/transactionUtils.ts:47]: Maybe `payload.call` needs to be converted to hex first? I'm fine with this staying as-is, but it seems redundant ;)

Comment (hbulgarini) [reply src/util/transactionUtils.ts:61]: >  I'm fine with this staying as-is, but it seems redundant ;)

I know ! i believe the same as you but for some reason that i don't understand why it does not work. This is the error i get:
`Error: createType(Call):: findMetaCall: Unable to find Call with index 0x9490/[148,144]`

File: src/types/transactionTypes.ts [hunk 13]
```diff
@@ -33,6 +33,11 @@ export enum TransactionStatusEnum {
   FAILED = 'FAILED'
 }
 
+export enum SwitchTabType {
```
Comment (goldsteinsveta) [inline src/types/transactionTypes.ts:36]: naming: SwitchTabEnum?

File: src/components/TransactionPayload.tsx [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+  return (
+    <Card elevation={transactionDisplayProps?.size === 'sm' ? 23 : 24} className={classes.card}>
+      {payloadHex && <TransactionHeader type={type} status={status} />}
+      {estimatedFeeLoading && (
+        <div className={classes.loader}>
+          <CircularProgress />
+        </div>
+      )}
+      {!estimatedFeeLoading && tab === SwitchTabType.PAYLOAD && payloadHex}
+      {!estimatedFeeLoading && tab === SwitchTabType.DECODED && transactionDisplayPayload && (
```
Comment (goldsteinsveta) [inline src/components/TransactionPayload.tsx:89]: just curious, why do you need this `&& transactionDisplayPayload` condition?

File: src/components/TransactionPayload.tsx [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+        <ReactJson src={transactionDisplayPayload} enableClipboard name={false} />
```
Comment (goldsteinsveta) [inline src/components/TransactionPayload.tsx:90]: typography nit: could you please wrap this in `<Typography variant="subtitle2" component="span"/>`. Thanks!

**Review by goldsteinsveta — APPROVED** [review]
This is beautiful!

File: src/types/transactionTypes.ts (hunk 13 above)
Comment (hbulgarini) [reply src/types/transactionTypes.ts:36]: sure!

File: src/components/TransactionPayload.tsx (hunk 14 above)
Comment (hbulgarini) [reply src/components/TransactionPayload.tsx:89]: It was there because of ReactJson component's src prop demands an object type and do not allow undefined.

File: src/components/TransactionPayload.tsx (hunk 15 above)
Comment (hbulgarini) [reply src/components/TransactionPayload.tsx:90]: of course!
